src/utils/playground.ts

```diff
@@ -0,0 +1,118 @@
+import * as vscode from 'vscode';
+import micromatch from 'micromatch';
+import * as fs from 'fs';
+import * as path from 'path';
+
+import { createLogger } from '../logging';
+
+const log = createLogger('playground utils');
+
+export class FileStat implements vscode.FileStat {
+  constructor(private fsStat: fs.Stats) {}
+
+  get type(): vscode.FileType {
+    if (this.fsStat.isFile()) {
+      return vscode.FileType.File;
+    }
+    if (this.fsStat.isDirectory()) {
+      return vscode.FileType.Directory;
+    }
+    if (this.fsStat.isSymbolicLink()) {
+      return vscode.FileType.SymbolicLink;
+    }
+
+    return vscode.FileType.Unknown;
+  }
+
+  get isFile(): boolean | undefined {
+    return this.fsStat.isFile();
+  }
+
+  get isDirectory(): boolean | undefined {
+    return this.fsStat.isDirectory();
+  }
+
+  get isSymbolicLink(): boolean | undefined {
+    return this.fsStat.isSymbolicLink();
+  }
+
+  get size(): number {
+    return this.fsStat.size;
+  }
+
+  get ctime(): number {
+    return this.fsStat.ctime.getTime();
+  }
+
+  get mtime(): number {
+    return this.fsStat.mtime.getTime();
+  }
+}
+
+const getFileNames = (filePath: string): Promise<string[]> => {
+  return fs.promises.readdir(filePath);
+};
+
+const getStat = async (filePath: string): Promise<vscode.FileStat> => {
+  return new FileStat(await stat(filePath));
+};
+
+const stat = (filePath: string): Promise<fs.Stats> => {
+  return fs.promises.stat(filePath);
```

Comment:
This should probably be:

```suggestion
  return fs.promises.lstat(filePath);
```

because otherwise you’d end up with an infinite loop inside `getPlaygrounds()` if you encounter a symbolic link that points to one of its parent directories. (If you do want to follow symbolic links, you probably need to keep track of which files you already visisted + use `fs.realpath` to avoid duplicates)